### ISSUE TITLE
(master) rdoc test suite failing with Java::JavaLang::ArrayIndexOutOfBoundsException

### ISSUE THREAD

**Issue by flavorjones** [jruby/jruby#8083]
The test suite for ruby/rdoc has been failing on JRuby master for the last two weeks or so. This suite passes with JRuby stable releases.

**Environment Information**

- `jruby 9.4.6.0-SNAPSHOT (3.1.4) 2024-01-22 7609e7b3ad OpenJDK 64-Bit Server VM 11.0.21+9 on 11.0.21+9 +jit [x86_64-linux]`
- github actions ubuntu-latest image

Last green test suite run was with `jruby 9.4.6.0-SNAPSHOT (3.1.4) 2024-01-20 e43e3b56ed` in https://github.com/ruby/rdoc/actions/runs/7613213302/job/20732599904.

First failing test suite run was with `jruby 9.4.6.0-SNAPSHOT (3.1.4) 2024-01-22 7609e7b3ad` in https://github.com/ruby/rdoc/actions/runs/7619307831/job/20752157542

**Expected Behavior**

I believe the test suite is valid and should pass.

**Actual Behavior**

See failure example at https://github.com/ruby/rdoc/actions/runs/7619307831/job/20752157542

```
Error: test_parse_desclist(TestRDocRdBlockParser): Java::JavaLang::ArrayIndexOutOfBoundsException: arraycopy: last source index 51 out of bounds for byte[45]
java.base/java.lang.System.arraycopy(Native Method)
org.jruby.dist/org.jruby.util.ByteList.<init>(ByteList.java:227)
org.jruby.ext.strscan.RubyStringScanner.newString(RubyStringScanner.java:877)
org.jruby.ext.strscan.RubyStringScanner.extractRange(RubyStringScanner.java:249)
org.jruby.ext.strscan.RubyStringScanner.rest(RubyStringScanner.java:685)
org.jruby.ext.strscan.RubyStringScanner$INVOKER$i$0$0$rest.call(RubyStringScanner$INVOKER$i$0$0$rest.gen)
org.jruby.dist/org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:193)
org.jruby.dist/org.jruby.ir.interpreter.InterpreterEngine.processCall(InterpreterEngine.java:[35](https://github.com/ruby/rdoc/actions/runs/7619307831/job/20752157542#step:6:36)0)
org.jruby.dist/org.jruby.ir.interpreter.StartupInterpreterEngine.interpret(StartupInterpreterEngine.java:66)
org.jruby.dist/org.jruby.ir.interpreter.InterpreterEngine.interpret(InterpreterEngine.java:76)
org.jruby.dist/org.jruby.internal.runtime.methods.MixedModeIRMethod.INTERPRET_METHOD(MixedModeIRMethod.java:164)
org.jruby.dist/org.jruby.internal.runtime.methods.MixedModeIRMethod.call(MixedModeIRMethod.java:151)
org.jruby.dist/org.jruby.internal.runtime.methods.DynamicMethod.call(DynamicMethod.java:212)
org.jruby.dist/org.jruby.runtime.callsite.CachingCallSite.cacheAndCall(CachingCallSite.java:[45](https://github.com/ruby/rdoc/actions/runs/7619307831/job/20752157542#step:6:46)6)
org.jruby.dist/org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:195)
com.headius.racc.Cparse$CparseParams.parse_main(Cparse.java:294)
com.headius.racc.Cparse$Parser.racc_cparse(Cparse.java:7[51](https://github.com/ruby/rdoc/actions/runs/7619307831/job/20752157542#step:6:52))
com.headius.racc.Cparse$Parser$INVOKER$i$2$0$racc_cparse.call(Cparse$Parser$INVOKER$i$2$0$racc_cparse.gen)
```

**Comment by headius** [comment]
Java exceptions leaking out to Ruby almost always indicate a bug. I'll look into it. Probably something I broke with the recent strscan update.

**Comment by headius** [comment]
I found the issue. Here's a simple reproduction:

```
$ jruby -rstrscan -e 'str = "hellohello"[5..-1]; StringScanner.new(str).scan(/hello/)'
Unhandled Java exception: java.lang.ArrayIndexOutOfBoundsException: arraycopy: last source index 15 out of bounds for byte[10]
java.lang.ArrayIndexOutOfBoundsException: arraycopy: last source index 15 out of bounds for byte[10]
            arraycopy at java/lang/System.java:-2
               <init> at org/jruby/util/ByteList.java:227
            newString at org/jruby/ext/strscan/RubyStringScanner.java:877
        extractBegLen at org/jruby/ext/strscan/RubyStringScanner.java:260
                 scan at org/jruby/ext/strscan/RubyStringScanner.java:342
                 scan at org/jruby/ext/strscan/RubyStringScanner.java:391
```

Fix coming.

**Comment by headius** [comment]
Fixed by ruby/strscan#87. Once that is released we can pull in the updated gem.

**Comment by headius** [comment]
Fixed strscan may need to be backported to 9.3, but that copy is missing several other features and new version breaks some tests.

**Comment by headius** [comment]
Backported to 9.3 in #8089.
